From 918a462ed993b5d8f1db3e3592a00210cc53cb5c Mon Sep 17 00:00:00 2001
From: Lubos Kardos <lkardos@redhat.com>
Date: Fri, 12 Jun 2015 17:26:25 +0200
Subject: [PATCH] Warn when macro is expanded in comment.

This warning is showed only when readLine() is called with
STRIP_COMMENTS. Basically it means that this warning is not showed
for macros in scriptlets (rhbz:#1224660).

(cherry picked from commit 2a3f49585e5bd82d0bbfe5b9d6cdf24d9501b5cd)
Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 build/parseSpec.c | 51 ++++++++++++++++++++++++++++++++++++++---------
 1 file changed, 42 insertions(+), 9 deletions(-)

diff --git a/build/parseSpec.c b/build/parseSpec.c
index 9b0c542cf7ac..39adedcd8638 100644
--- a/build/parseSpec.c
+++ b/build/parseSpec.c
@@ -151,8 +151,45 @@ static int restoreFirstChar(rpmSpec spec)
     return 0;
 }
 
+static int expandMacrosInSpecBuf(rpmSpec spec, int strip)
+{
+    char *lbuf = NULL;
+    int rc = 0, isComment = 0;
+
+     /* Don't expand macros (eg. %define) in false branch of %if clause */
+    if (!spec->readStack->reading)
+	return 0;
+
+    lbuf = spec->lbuf;
+    SKIPSPACE(lbuf);
+    if (lbuf[0] == '#')
+	isComment = 1;
+
+    lbuf = xstrdup(spec->lbuf);
+
+    rc = expandMacros(spec, spec->macros, spec->lbuf, spec->lbufSize);
+    if (rc) {
+	rpmlog(RPMLOG_ERR, _("line %d: %s\n"),
+		spec->lineNum, spec->lbuf);
+	goto exit;
+    }
+
+    if (strip & STRIP_COMMENTS &&
+	isComment && !rstreq(spec->lbuf, lbuf)) {
+
+	rpmlog(RPMLOG_WARNING,
+	    _("Macro expanded in comment on line %d: %s\n"),
+	    spec->lineNum, lbuf);
+    }
+
+exit:
+    free(lbuf);
+
+    return rc;
+}
+
 /* Return zero on success, 1 if we need to read more and -1 on errors. */
-static int copyNextLineFromOFI(rpmSpec spec, OFI_t *ofi)
+static int copyNextLineFromOFI(rpmSpec spec, OFI_t *ofi, int strip)
 {
     /* Expand next line from file into line buffer */
     if (!(spec->nextline && *spec->nextline)) {
@@ -203,13 +240,9 @@ static int copyNextLineFromOFI(rpmSpec spec, OFI_t *ofi)
 	}
 	spec->lbufOff = 0;
 
-	/* Don't expand macros (eg. %define) in false branch of %if clause */
-	if (spec->readStack->reading &&
-	    expandMacros(spec, spec->macros, spec->lbuf, spec->lbufSize)) {
-		rpmlog(RPMLOG_ERR, _("line %d: %s\n"),
-			spec->lineNum, spec->lbuf);
-		return -1;
-	}
+	if (expandMacrosInSpecBuf(spec, strip))
+	    return -1;
+
 	spec->nextline = spec->lbuf;
     }
     return 0;
@@ -319,7 +352,7 @@ int readLine(rpmSpec spec, int strip)
 	ofi = spec->fileStack;
 
 	/* Copy next file line into the spec line buffer */
-	rc = copyNextLineFromOFI(spec, ofi);
+	rc = copyNextLineFromOFI(spec, ofi, strip);
 	if (rc > 0) {
 	    if (startLine == 0)
 		startLine = spec->lineNum;
